spacy_llm/pipeline/llm.py

```diff
@@ -126,25 +134,42 @@ def pipe(self, stream: Iterable[Doc], *, batch_size: int = 128) -> Iterator[Doc]
         batch_size (int): The number of documents to buffer.
         YIELDS (Doc): Processed documents in order.
         """
+        if not Doc.has_extension("llm_io"):
+            Doc.set_extension("llm_io", default=defaultdict(dict))
+
         error_handler = self.get_error_handler()
         for doc_batch in spacy.util.minibatch(stream, batch_size):
             is_cached = [doc in self._cache for doc in doc_batch]
             noncached_doc_batch = [
                 doc for doc, cached_doc in zip(doc_batch, is_cached) if not cached_doc
             ]
             try:
-                prompts = self._task.generate_prompts(noncached_doc_batch)
-                responses = self._backend(prompts)
+                # We need to use tee to store generated prompts/responses
+                # Otherwise, they will be consumed
+                prompts = tee(self._task.generate_prompts(noncached_doc_batch))
+                responses = tee(self._backend(prompts))
```

Comment:
Not with this naming, but what I meant was more like:

```suggestion
                prompts = tee(self._task.generate_prompts(noncached_doc_batch))
                if self._save_io:
                    prompts, prompts_for_saving = tee(prompts)
                responses = self._backend(prompts))
                if self._save_io:
                    responses, responses_for_saving = tee(responses)
```